hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,480 @@
+          nextBatchProcessTimeMs += batchIntervalMs;
+        } else {
+          // Otherwise, wait for batchIntervalMs based on the current timestamp
+          nextBatchProcessTimeMs = currTimeMs + batchIntervalMs;
+        }
+
+        long waitMs = nextBatchProcessTimeMs - System.currentTimeMillis();
+        executorService.schedule(
+            new BatchCreateMarkerRunnable(), Math.max(0L, waitMs), TimeUnit.MILLISECONDS);
+        LOG.info("Wait for " + waitMs + " ms, next batch time: " + nextBatchProcessTimeMs);
+      }
+    }
+    return future;
+  }
+
+  /**
+   * Deletes markers in the directory.
+   *
+   * @param markerDir marker directory path
+   * @return {@code true} if successful; {@code false} otherwise.
+   */
+  public Boolean deleteMarkers(String markerDir) {
+    Path markerDirPath = new Path(markerDir);
+    try {
+      if (fileSystem.exists(markerDirPath)) {
+        FileStatus[] fileStatuses = fileSystem.listStatus(markerDirPath);
+        List<String> markerDirSubPaths = Arrays.stream(fileStatuses)
+            .map(fileStatus -> fileStatus.getPath().toString())
+            .collect(Collectors.toList());
+
+        if (markerDirSubPaths.size() > 0) {
+          SerializableConfiguration conf = new SerializableConfiguration(fileSystem.getConf());
+          int actualParallelism = Math.min(markerDirSubPaths.size(), parallelism);
+          hoodieEngineContext.foreach(markerDirSubPaths, subPathStr -> {
+            Path subPath = new Path(subPathStr);
+            FileSystem fileSystem = subPath.getFileSystem(conf.get());
+            fileSystem.delete(subPath, true);
+          }, actualParallelism);
+        }
+
+        boolean result = fileSystem.delete(markerDirPath, true);
+        LOG.info("Removing marker directory at " + markerDirPath);
+        return result;
+      }
+    } catch (IOException ioe) {
+      throw new HoodieIOException(ioe.getMessage(), ioe);
+    }
+    return false;
+  }
+
+  /**
+   * Gets the marker file index from the marker file path.
+   *
+   * E.g., if the marker file path is /tmp/table/.hoodie/.temp/000/MARKERS3, the index returned is 3.
+   *
+   * @param markerFilePathStr full path of marker file
+   * @return the marker file index
+   */
+  private int getMarkerFileIndex(String markerFilePathStr) {
+    String markerFileName = new Path(markerFilePathStr).getName();
+    int prefixIndex = markerFileName.indexOf(MARKERS_FILENAME_PREFIX);
+    if (prefixIndex < 0) {
+      return -1;
+    }
+    try {
+      return Integer.parseInt(markerFileName.substring(prefixIndex + MARKERS_FILENAME_PREFIX.length()));
+    } catch (NumberFormatException nfe) {
+      nfe.printStackTrace();
+    }
+    return -1;
+  }
+
+  /**
+   * Syncs all markers maintained in the marker files from the marker directory in the file system.
+   *
+   * @param markerDir marker directory path
+   */
+  private void syncAllMarkersFromFile(String markerDir) {
```

Comment:
@vinothchandar @nsivabalan  I created this method for syncing the markers from marker files to the in-memory cache in case of timeline service failure.  Any suggestion on how we use this?  When the timeline service is initialized, it does not know the new commit instant thus the marker file directory.  So I'm not sure if we should scan the whole hoodie temporary folder (.hoodie/.temp), find all marker directories and sync them.

Reply:
yeah, I could not think of other ways. will see if vinoth has any suggestions

Reply:
This is now resolved with lazy sync.